### ISSUE TITLE
How to Use Custom Labels

### ISSUE THREAD

**Issue by yuzhenmao** [pytorch/torchtune#2111]
Hi, I recently went through the tutorial for [Fine-tune LLama 3 with Chat Data](https://pytorch.org/torchtune/0.2/tutorials/chat.html) and [chat_dataset](https://pytorch.org/torchtune/stable/generated/torchtune.datasets.chat_dataset.html#torchtune.datasets.chat_dataset). However, I couldn’t find a solution for setting custom labels.

Currently, it seems Torchtune only supports fine-tuning where labels are either all assistant output (`train_on_input`=False) or include the prompt (`train_on_input`=True). What if I want to use only a subset of the assistant output as labels?

Is there a straightforward way to modify the code to support this? I’d appreciate your guidance.

**Comment by calvinpelletier** [comment]
Hi @yuzhenmao , do you mean a subset of the assistant messages, or a subset of the output within a single assistant message?

If you would like to train on a subset of the assistant output, you'll need to write a custom message transform.

At a high level, you'll need a dataset builder. This is a function that takes a `torchtune.modules.tokenizers.ModelTokenizer` as the first argument and returns a `torch.utils.data.Dataset`:

```python
def custom_dataset(tokenizer: ModelTokenizer) -> Dataset:
    ...
```

You'll point to this builder in your config (any additional key-value pairs in the `dataset` config will be passed to your builder as kwargs): 

```yaml
dataset:
  _component_: importpath.to.your.custom_dataset
```

If your data is compatible with [huggingface's `load_dataset`](https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path), then you can use the `torchtune.datasets._sft.SFTDataset` class with a custom message transform. Your builder will look similar to this:

```python
def custom_dataset(
    tokenizer: ModelTokenizer,
    *,
    source: str,
    packed: bool = False,
    filter_fn: Optional[Callable] = None,
    split: str = "train",
    **load_dataset_kwargs: Dict[str, Any],
) -> Union[SFTDataset, PackedDataset]:
    message_transform = CustomMessageTransform()
    ds = SFTDataset(
        source=source,
        message_transform=message_transform,
        model_transform=tokenizer,
        split=split,
        filter_fn=filter_fn,
        **load_dataset_kwargs,
    )
    if packed:
        if tokenizer.max_seq_len is None:
            raise ValueError(
                "PackedDataset requires a max_seq_len to be set on the tokenizer."
            )
        return PackedDataset(ds, max_seq_len=tokenizer.max_seq_len)
    return ds
```

The message transform is a callable that takes in a single row from your dataset and returns a list of `torchtune.data._messages.Message` objects. The `masked` property of each `Message` object is how you can train on only a subset of the assistant messages. If `message.masked==True`, then it will not be used in the loss calculation. [Here's an example.](https://github.com/pytorch/torchtune/blob/5eb04cd934ad84efff61e5dbf7a054fd7af184ec/torchtune/data/_messages.py#L616)

However, if you want to train on a subset of the output within a single assistant message, then it's a bit more complicated. If this is the case, let me know and I can help with that.

Let me know if you have any questions!

**Comment by yuzhenmao** [comment]
That is exactly what I need! Thank you @calvinpelletier so much for your detailed reply!
